cyclepainter.py

```diff
@@ -15,23 +15,56 @@
 
 
 def dist(p1, p2):
-    ''' UTILITY method for euclidean distance'''
+    ''' UTILITY method for euclidean distance '''
     return ((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2)**0.5
 
 def ccw(a, b, c):
     ''' UTILITY method for determining orientation '''
+    # Algorithm copied from https://bryceboe.com/2006/10/23/line-segment-intersection-algorithm/
+    # Recall that a Riemann surface has a canonical orientation given by w=dx^dy if z=x+iy is the complex coordinate
+    # ccw returns True=1 if w(a->b,a->c)>0, i.e True if the vectors ab, ac are positively oriented.
+    # When a,b,c are colinear, we can no longer ask about the orientation as it is not defined. 
+    # Note ccw as a function will not detect this problem on its own. 
     return bool((c[1] - a[1]) * (b[0] - a[0]) > (b[1] - a[1]) * (c[0] - a[0]))
 
+def collinear(a, b, c):
```

Comment:
This equality can still be treacherous in some cases due to limitations of the floating point arithmetic. 
```
>>> collinear((0,0), (2**0.5, 3**0.5), (7 * 2**0.5, 7 * 3**0.5))
False
```
I do not have a good solution - but I guess it's very improbable that this would affect Cyclepainter's workings

Reply:
Yep this is definitely an issue, one I agree is unlikely to have an effect as this is a false negative caused by approximations to square roots. The case we are thinking about is when a segment crosses a branch cut, and cyclepainter will ensure we don't get these non-zero but tiny floats by how the assignment works I believe (thinking about CyclePainterPath where the intersection is calculated by the method, and one can verify that sage will necessarily set this to 0).